### ISSUE TITLE
Descriptor type variable in generic incorrectly expects __set__ value type instead of descriptor type itself

### ISSUE THREAD

**Issue by max-marek** [astral-sh/ty#4287]
### Summary

When a dataclass is generic over a type variable bounded to a descriptor class, the type checker seems to incorrectly require the constructor argument to match the value parameter type of the descriptor’s `__set__` method, rather than the descriptor type itself.

**Minimal reproducible example:**

```python
from dataclasses import dataclass

class SomeDescriptor:
    def __set__(self, parent: object, value: int) -> None:
        ...

@dataclass
class SomeParent[SomeDescriptorType: SomeDescriptor]:
    attr: SomeDescriptorType

SomeParent(SomeDescriptor())
```

**Error:**

```
Argument is incorrect: Expected `int`, found `SomeDescriptor`  ty[invalid-argument-type]
```

The argument to SomeParent should be of type SomeDescriptorType (i.e., SomeDescriptor), so the call SomeParent(SomeDescriptor()) should type‑check without errors.

### Version

ty 0.0.72

**Comment by carljm** [comment]
Interesting case, thanks for the report.

If we simplify the scenario to remove the bounded typevar and just use the `Descriptor` type directly as the field type:

```py
from dataclasses import dataclass
from typing import reveal_type

class Descriptor:
    def __set__(self, parent: object, value: int) -> None: ...

@dataclass
class Example:
    attr: Descriptor

Example(Descriptor())
Example(1)
```

pyright, ty, and pyrefly all treat this field as a descriptor (that is, accept `int` and reject `Descriptor`). Mypy and zuban treat it as a regular field (accept `Descriptor`, reject `int`).

If you change the field definition to `attr: Descriptor = Descriptor()`, ty/pyrefly/pyright behavior remains unchanged; mypy/zuban switch to treating the field as a descriptor.

I think the mypy/zuban behavior is better here -- if there is no descriptor assigned on the class, then at runtime there is no descriptor behavior.

If we switch to the bounded typevar, all non-ty type-checkers always treat as "not a descriptor". I don't think this difference in behavior is justified -- a typevar with upper bound of a descriptor type should be modeled as a descriptor just as much as a direct descriptor type.

So I think ty is correct to treat the bounded typevar case the same as the direct case, but I think ty is wrong to model descriptor behavior when there is no actual descriptor assigned on the class.

**Comment by carljm** [comment]
Hmm, unfortunately I think this intersects with what we used to call "the SQLAlchemy problem". We used to model the runtime behavior more closely (in terms of requiring a class-level assignment before considering an annotated attribute to behave as a descriptor) and we had to roll that back because some libraries (including SQLAlchemy) relied on annotation-only attributes behaving as descriptors. More historical context in #384 and linked issues.

So that's the origin of our current (apparently inconsistent with runtime) behavior here. Which leaves it unclear to me what behavior we should change here, without reintroducing breakage for SQLAlchemy et al.

**Comment by carljm** [comment]
Mypy specifically handles this differently for stdlib dataclasses vs `dataclass_transform` classes, and that difference was specifically motivated by SQLAlchemy, in https://github.com/python/mypy/issues/14868. That might be a defensible distinction to make (though in general I don't like introducing differences between stdlib dataclass and third-party `dataclass_transform`), considering that for stdlib dataclasses we can basically be sure the descriptor won't magically appear on the class in some other way, whereas for `dataclass_transform` we unfortunately (thanks to SQLAlchemy) have to assume that it might.

**Comment by carljm** [comment]
Putting this in stable milestone since (for stdlib dataclasses) we get this bound-typevar case wrong relative to all other type-checkers.
